<div style="display: flex; width: 100%;">
    <div style="flex: 1; padding: 0px;">
        <p>© Albert Palacios Jiménez, 2023</p>
    </div>
    <div style="flex: 1; padding: 0px; text-align: right;">
        <img src="../assets/ieti.png" height="32" alt="Logo de IETI" style="max-height: 32px;">
    </div>
</div>
<hr/>

# Dibuix

Molts llenguatges de programació tenen eines o llibreries, per obrir finestres, fer-hi dibuixos o mostrar-hi imatges.

Per poder fer anar aquestes eines, calen coneixements bàsics de programació: variables, funcions, condicions, bucles, ...

En Python hi ha moltes llibreries, però una de les més senzilles és **pyGame**.

Instal·lació:
```bash
pip install pygame
pip install ipykernel -U --user --force-reinstall
```

A macOS i linux amb *brew*:
```bash
python3 -m pip install pygame --break-system-package
python3 -m pip install ipykernel -U --user --force-reinstall --break-system-package
```

## Finestres i aplicacions

Per poder dibuixar, primer necessitem una finestra d'aplicació.

```python
screen = pygame.display.set_mode((640, 480))
pygame.display.set_caption('Window Title')
```

Però un cop tenim la finestra, s'ha de definir què fer amb ella. És a dir, el **bucle de l'aplicació**

```python
def main():
    is_looping = True

    while is_looping:
        is_looping = app_events()
        app_run()
        app_draw()

        clock.tick(30) # Limitar a 30 FPS

    # Fora del bucle, tancar l'aplicació
    pygame.quit()
    sys.exit()
```

### Exemple 000

In [4]:
# Codi de l'exemple
import utils; utils.show_code('./exemple000.py')

```python
import math
import os
os.environ['PYGAME_HIDE_SUPPORT_PROMPT'] = "hide"
import pygame
import sys

# Definir colors
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE  = (0, 0, 255)
PURPLE = (128, 0, 128)
ORANGE = (255, 165, 0) 

pygame.init()
clock = pygame.time.Clock()

# Definir la finestra
screen = pygame.display.set_mode((640, 480))
pygame.display.set_caption('Window Title')

# Bucle de l'aplicació
def main():
    is_looping = True

    while is_looping:
        is_looping = app_events()
        app_run()
        app_draw()

        clock.tick(60) # Limitar a 60 FPS

    # Fora del bucle, tancar l'aplicació
    pygame.quit()
    sys.exit()

# Gestionar events
def app_events():
    for event in pygame.event.get():
        if event.type == pygame.QUIT: # Botó tancar finestra
            return False
    return True

# Fer càlculs
def app_run():
    pass

# Dibuixar
def app_draw():
    
    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Escriure un text de prova
    font = pygame.font.SysFont("Arial", 55)
    text = font.render('Hello World!', True, BLACK)
    screen.blit(text, (50, 50))

    # Actualitzar el dibuix a la finestra
    pygame.display.update()

if __name__ == "__main__":
    main()
```

In [6]:
# Fer anar l'exemple
import utils; utils.run_code('./exemple000.py')

# Dibuix

Un cop tenim definida la finesta, i el bucle de l'aplciació, ja es poden dibuixar cosses al a funció **app_draw**

```python
def app_draw():
    
    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Escriure un text de prova
    font = pygame.font.SysFont("Arial", 55)
    text = font.render('Hello World!', True, BLACK)
    screen.blit(text, (50, 200))

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

## Ordre de dibuix (canvas)

Normalment les llibreries de dibuix funcionen com un quadre *(canvas)*, són una superfície a la que vas dibuixant. I el què dibuixes últim, queda per damunt de tot el què ja has dibuixat.

<br/>
<center><img src="./assets/drawingorder.png" style="max-height: 400px" alt="">
<br/></center>
<br/>
<br/>

## Eix de coordenades

Indiquem a quina posició volem dibuixar a través de coordenades X,Y.

A **pygame** les coordenades són **"top/left"**, el què vol dir que el 0,0 de la finestra és la posició de dalt a l'equerra.

A l'exemple de la imatge, les coordenades són (x = 80, y = 50):

<br/>
<center><img src="./assets/coordinates.png" style="max-height: 250px" alt="">
<br/></center>
<br/>
<br/>

## Dibuix basic

Pygame proporciona diverses funcions per dibuixar formes bàsiques:
```python
# Una línia entre dos punts.
pygame.draw.line(screen, color, start_pos, end_pos, width)

# Un rectangle 
pygame.draw.rect(screen, color, pygame.Rect(x, y, width, height))

# Un cercle
pygame.draw.circle(screen, GREEN, (x, y), radius)

# Una el·lipse dins d'un rectangle delimitador.
pygame.draw.ellipse(screen, color, pygame.Rect(x, y, width, height))

# Un arc dins d'un rectangle delimitador.
pygame.draw.arc(screen, color, pygame.Rect(x, y, width, height), start_angle, end_angle, width)

# Un polígon connectant una sèrie de punts.
pygame.draw.polygon(screen, color, [(x1, y1), (x2, y2), (x3, y3)], width)

# Una sèrie de línies connectades entre diversos punts.
pygame.draw.lines(screen, color, closed, point_list, width)
```


### Exemple 001

Dibuix de formes bàsiques, fixar-se en com s'escriuen les coordenades i en l'ordre de dibuix.

In [9]:
# Codi de l'exemple
import utils; utils.show_code('./exemple001.py', function_name='app_draw')

```python
def app_draw():
    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella de coordenades (llibreria utils)
    # utils.draw_grid(pygame, screen, 50)

    # Dibuixar un rectangle vermell a la posició (150, 200) de mida (75, 100)
    pygame.draw.rect(screen, RED, pygame.Rect(150, 200, 75, 100))

    # Dibuixar un cercle verd a la posició (100, 150) de radi 10
    pygame.draw.circle(screen, GREEN, (100, 150), 50)

    # Dibuixar una linia des de l'origen del rectangle fins al centre del cercle
    pygame.draw.line(screen, BLUE, (150, 200), (100, 150), 5)

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

In [10]:
# Fer anar el codi d'exemple
import utils; utils.run_code('./exemple001.py')

### Exercici 000

Modifica la funció **app_draw** de l'arxiu **"./exercici000.py"** per fer el següent dibuix, a partir de dos bucles **for**

**Nota**: La mida de gruix de la línia és 5

<br/>
<center><img src="./assets/exercici000.png" style="max-height: 400px" alt="">
<br/></center>
<br/>

In [16]:
# Fer anar el codi de l'exercici
import utils; utils.run_code('./exercici000.py')

## Colors

**pyGame** permet fer dibuixos amb colors RGB (Red, Green Blue)

Els colors són **tuples** de tres paràmetres entre 0 i 255, on:

- El primer valor representa el vermell (Red)
- El segon valor representa el verd (Green)
- El tercer valor representa el blau (Blue)

Així, els colors clàssics són:

<span style="color:rgb(255, 255, 255)">■</span> WHITE = (255, 255, 255)  
<span style="color:rgb(0, 0, 0)">■</span> BLACK = (0, 0, 0)  
<span style="color:rgb(255, 0, 0)">■</span> RED = (255, 0, 0)  
<span style="color:rgb(0, 255, 0)">■</span> GREEN = (0, 255, 0)  
<span style="color:rgb(0, 0, 255)">■</span> BLUE = (0, 0, 255)  
<span style="color:rgb(255, 255, 0)">■</span> YELLOW = (255, 255, 0)  
<span style="color:rgb(0, 255, 255)">■</span> CYAN = (0, 255, 255)  
<span style="color:rgb(255, 0, 255)">■</span> MAGENTA = (255, 0, 255)  
<span style="color:rgb(255, 165, 0)">■</span> ORANGE = (255, 165, 0)  
<span style="color:rgb(128, 0, 128)">■</span> PURPLE = (128, 0, 128)  
<span style="color:rgb(255, 192, 203)">■</span> PINK = (255, 192, 203)  
<span style="color:rgb(128, 128, 128)">■</span> GRAY = (128, 128, 128)  
<span style="color:rgb(165, 42, 42)">■</span> BROWN = (165, 42, 42)  
<span style="color:rgb(0, 0, 128)">■</span> NAVY = (0, 0, 128)  
<span style="color:rgb(0, 128, 128)">■</span> TEAL = (0, 128, 128)  
<span style="color:rgb(0, 255, 0)">■</span> LIME = (0, 255, 0)  
<span style="color:rgb(75, 0, 130)">■</span> INDIGO = (75, 0, 130)  
<span style="color:rgb(238, 130, 238)">■</span> VIOLET = (238, 130, 238)


## Relleus i Emplenats

Hi ha dos modes de dibuix:

- Emplenat (**fill**): És el color o patró que omple l’interior d'una forma. Per exemple, si dibuixes un cercle i l'emplenes de color blau, tot l'interior del cercle serà blau.

- Relleu (**stroke**): És el contorn o línia exterior de la forma, que pot tenir un color i gruix diferents. Si dibuixes un cercle amb relleu vermell de 5 píxels, el contorn del cercle serà vermell amb una amplada de 5 píxels, però l'interior pot estar buit o emplenat d’un altre color.

Aquestes propietats permeten que una forma tingui un aspecte amb un interior (emplenat) diferent del seu contorn (relleu)

In [17]:
# Codi de l'exemple
import utils; utils.show_code('./exemple002.py', function_name='app_draw')

```python
def app_draw():
    
    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella
    utils.draw_grid(pygame, screen, 50)

    # Primer quadrat: emplenat taronja amb relleu blau
    pygame.draw.rect(screen, ORANGE, (50, 50, 100, 100))  # Emplenat
    pygame.draw.rect(screen, BLUE, (50, 50, 100, 100), 5)  # Relleu

    # Segon quadrat: només emplenat taronja
    pygame.draw.rect(screen, ORANGE, (200, 50, 100, 100))  # Emplenat

    # Tercer quadrat: només relleu blau
    pygame.draw.rect(screen, BLUE, (350, 50, 100, 100), 5)  # Relleu

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

In [18]:
# Fer anar l'exemple
import utils; utils.run_code('./exemple002.py')

### Exercici 001

Al codi **"exercici001.py"**, modifica la funció **app_draw** per fer el següent dibuix:

**Nota**: Per dibuixar l'arc taronja, fer servir la següent informació:

- Coordenades: x = 400, y = 250
- Mida: ample = 200, alt = 100
- Angle inicial: math.radians(45)
- Angle final: math.radians(180)
- Gruix: 5

<br/>
<center><img src="./assets/exercici001.png" style="max-height: 400px" alt="">
<br/></center>
<br/>

In [23]:
# Fer anar el codi de l'exercici
import utils; utils.run_code('./exercici001.py')

## Texts

Per dibuixar un text cal seguir els següents passos:

- Definir un estil de font
- Definir un text (si es pot, fora del bucle)
- Dibuixar el text (dins del bucle)

**Nota:** Com que l'estil de font, i el text són variables, millor posar-les fora del bucle de l'aplicació.

## Imatges

Per dibuixar imatges cal

- Carregar la imatge en una variable
- Dibuixar la imatge (dins del bucle)

**Nota**: Quan es carrega la imatge se li pot canviar la mida, amb *pygame.transform.scale*

In [25]:
# Codi de l'exemple
import utils; utils.show_code('./exemple003.py')

```python
import math
import os
os.environ['PYGAME_HIDE_SUPPORT_PROMPT'] = "hide"
import pygame
import sys
import utils

# Definir colors
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE  = (0, 0, 255)
PURPLE = (128, 0, 128)
ORANGE = (255, 165, 0)  

pygame.init()
clock = pygame.time.Clock()

# Definir la finestra
screen = pygame.display.set_mode((640, 480))
pygame.display.set_caption('Window Title')

# Bucle de l'aplicació
def main():
    global font, text, image

    is_looping = True

    # Definir el text
    font = pygame.font.SysFont("Arial", 55)
    text = font.render('Hello Arial!', True, BLACK)

    # Carregar la imatge
    image = pygame.image.load('./assets/exemple003.png')
    image = scale_image(image, target_width=100)

    while is_looping:
        is_looping = app_events()
        app_run()
        app_draw()

        clock.tick(60) # Limitar a 60 FPS

    # Fora del bucle, tancar l'aplicació
    pygame.quit()
    sys.exit()

# Gestionar events
def app_events():
    for event in pygame.event.get():
        if event.type == pygame.QUIT: # Botó tancar finestra
            return False
    return True

# Fer càlculs
def app_run():
    pass

# Dibuixar
def app_draw():
    
    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella
    utils.draw_grid(pygame, screen, 50)

    # Text
    screen.blit(text, (50, 50))

    # Imatge
    screen.blit(image, (400, 50))

    # Actualitzar el dibuix a la finestra
    pygame.display.update()

def scale_image(image, target_width=None, target_height=None):

    original_width, original_height = image.get_size()
    aspect_ratio = original_height / original_width

    if target_width and not target_height:  # Escalar per ample mantenint la proporció
        new_width = target_width
        new_height = int(target_width * aspect_ratio)
    elif target_height and not target_width:  # Escalar per altura mantenint la proporció
        new_height = target_height
        new_width = int(target_height / aspect_ratio)
    elif target_width and target_height:  # Escalar deformant la imatge
        new_width = target_width
        new_height = target_height
    else:
        raise ValueError("Especifica almenys un dels dos paràmetres: target_width o target_height.")

    scaled_image = pygame.transform.scale(image, (new_width, new_height))
    return scaled_image

if __name__ == "__main__":
    main()
```

In [26]:
# Fer anar l'exemple
import utils; utils.run_code('./exemple003.py')

## Events, lògica i dibuix

Cal entendre que quan fem aplicacions gràfiques, el bucle de l'aplicació es divideix en parts:

- **Obtenció d'events**: és a dir la informació que rep l'aplicació

- **Càlculs lògics**: els que defineixen l'estat de l'aplicació, i modifiquen les variables

- **Dibuix**: a partir dels càlculs anteriors dibuixar el què correspon

**Important**: Separar la lògica dels càlculs de les instruccions de dibuix, permet que les instruccions que s'envien a la **tarjeta gràfica** estiguin més optimitzades, i l'aplicació funcioni més ràpid. 

Quan es posa codi per calcular la lògica dins de la funció de dibuix, les instruccions que s'envien a la tarja gràfica queden fragmentades, i no es poden optimitzar.

## Animacions 

Les animacions permeten canviar el valor d'una variable en el temps, per tal de fer efectes de moviment o de canvi d'estat.

Per tal que les animacions siguin coherents en tots els equips (sempre triguin el mateix), no canviarem els valors de les animacions en funció dels cops que hem dibuixat sinó **en funció del temps** que triquem a dibuixar.

In [27]:
# Codi de l'exemple (comprovar si l'usuari vol tancar l'aplicació per sortir del bucle)
import utils; utils.show_code('./exemple004.py', function_name='app_run')

```python
def app_run():
    global pos_x, dir_x

    delta_time = clock.get_time() / 1000.0  # Convertir a segons
    
    speed = 50  # píxels per segon
    displacement = speed * delta_time

    if (dir_x == "right"):
        pos_x = pos_x + displacement
        if (pos_x > 200):
            dir_x = "left"
    else:
        pos_x = pos_x - displacement
        if (pos_x < 100):
            dir_x = "right"
```

In [28]:
# Codi de l'exemple (comprovar si l'usuari vol tancar l'aplicació per sortir del bucle)
import utils; utils.show_code('./exemple004.py', function_name='app_draw')

```python
def app_draw():
    global pos_x, dir_x

    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella
    utils.draw_grid(pygame, screen, 50)

    # Draw
    pygame.draw.line(screen, BLUE, (pos_x, 100), (pos_x, 200), 5)

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

In [29]:
# Codi de l'exemple (canviar el valor de la variable posició X)
import utils; utils.run_code('./exemple004.py')

### Exercici 002

La següent funció, dona una posició X,Y al perímetre d'un cercle a partir dels paràmetres:

- Centre del cercle (x, y)
- Radi del cercle
- Angle (en graus)

In [ ]:
import math

def posicio_perimetre_cercle(center, radi, angle_graus):
    angle_radians = math.radians(angle_graus)  # Convertir l'angle a radians
    x = center[0] + radi * math.cos(angle_radians)    # Coordenada X
    y = center[1] + radi * math.sin(angle_radians)    # Coordenada Y
    return x, y

Al codi **"exercici002.py"**, fes l'animació d'una linia que dóna voltes relatives a un cercle:

**Nota**: Fes que la velocitat a la que canvia el valor de l'angle sigui de *50 * delta_time*

<center>
<video width="100%" controls allowfullscreen style="max-width: 90%; width: 400px; max-height: 200px">
  <source src="./assets/exercici002.mov" type="video/mp4">
</video>
</center>
<br/>

In [32]:
# Fer anar el codi de l'exercici
import utils; utils.run_code('./exercici002.py')

## Events de teclat

Els events són **missatges que el programa rep quan passa alguna cosa**, normalment quan l'usuari interacciona amb l'aplicació a través del mouse o el teclat.

Hi ha molts tipus d'events: de la finestra, del mouse, del teclat, ...

Amb la informació dels events, podem modificar variables del nostre codi (i conseqüentment modificar el dibuix)

### Exemple 005

Els **events de teclat** ens avisen quan s'apreta o s'allibera una tecla. Cal tenir en compte:

- Múltiples tecles poden estar apretades al mateix temps
- Hi ha tecles de modificació com *control*, *alt*, *command* ... que canvien el comportament de les tecles

No s'ha de confondre l'**event d'apretar o aixecar** una tecla:

- pygame.KEYDOWN
- pygame.KEYUP

Amb les **tecles** 'amunt' i 'avall'

- pygame.K_UP
- pygame.K_DOWN

In [35]:
# Codi de l'exemple
import utils; utils.show_code('./exemple005.py', function_name='app_events')

```python
def app_events():
    global dir_x, dir_y

    for event in pygame.event.get():
        if event.type == pygame.QUIT: # Botó tancar finestra
            return False
        elif event.type == pygame.KEYDOWN:  # Tecla premuda
            if event.key == pygame.K_UP:
                dir_y = 'up'
            elif event.key == pygame.K_DOWN:
                dir_y = 'down'
            elif event.key == pygame.K_LEFT:
                dir_x = 'left'
            elif event.key == pygame.K_RIGHT:
                dir_x = 'right'
        elif event.type == pygame.KEYUP:  # Tecla alliberada
            if event.key == pygame.K_UP:
                if dir_y == 'up':
                    dir_y = 'none'
            elif event.key == pygame.K_DOWN:
                if dir_y == 'down':
                    dir_y = 'none'
            elif event.key == pygame.K_LEFT:
                if dir_x == 'left':
                    dir_x = 'none'
            elif event.key == pygame.K_RIGHT:
                if dir_x == 'right':
                    dir_x = 'none'
    return True
```

In [36]:
# Codi de l'exemple
import utils; utils.show_code('./exemple005.py', function_name='app_run')

```python
def app_run():
    global dir_x, dir_y, pos_x, pos_y

    delta_time = clock.get_time() / 1000.0  # Convertir a segons
    
    speed = 50  # píxels per segon
    displacement = speed * delta_time

    if (dir_x == "right"):
        pos_x = pos_x + displacement
        if (pos_x > 200):
            pos_x = 200
    elif (dir_x == "left"):
        pos_x = pos_x - displacement
        if (pos_x < 100):
            pos_x = 100

    if (dir_y == "down"):
        pos_y = pos_y + displacement
        if (pos_y > 200):
            pos_y = 200
    elif (dir_y == "up"):
        pos_y = pos_y - displacement
        if (pos_y < 100):
            pos_y = 100
```

In [40]:
# Codi de l'exemple
import utils; utils.show_code('./exemple005.py', function_name='app_draw')

```python
def app_draw():
    global pos_x, pos_y

    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella
    utils.draw_grid(pygame, screen, 50)

    # Draw limits
    pygame.draw.rect(screen, BLUE, pygame.Rect(100, 100, 100, 100), 2)

    # Draw moving object
    pygame.draw.rect(screen, ORANGE, pygame.Rect(pos_x, pos_y, 15, 15))

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

In [41]:
# Fer anar l'exemple
import utils; utils.run_code('./exemple005.py')

### Exercici 003

Al codi **"exercici003.py"**, arregla l'exemple anterior perquè el quadre taronja no surti mai dels limits del quadre blau.

<br/>
<center><img src="./assets/exercici003.png" style="max-height: 200px;" alt="">
<br/></center>
<br/>

In [42]:
# Fer anar el codi de l'exercici
import utils; utils.run_code('./exercici003.py')

## Events de ratolí

Al igual que el teclat els events de ratolí ens avisen de:

- La posició del mouse
- Si s'apreta algún botó del ratolí
- Si s'allibera algún botó del ratolí

In [44]:
# Codi de l'exemple
import utils; utils.show_code('./exemple006.py', function_name='app_events')

```python
def app_events():
    global mouse_pos
    mouse_inside = pygame.mouse.get_focused() # El ratolí està dins de la finestra?

    for event in pygame.event.get():
        if event.type == pygame.QUIT: # Botó tancar finestra
            return False
        elif event.type == pygame.MOUSEMOTION:
            if mouse_inside:
                mouse_pos["x"] = event.pos[0]
                mouse_pos["y"] = event.pos[1]
            else:
                mouse_pos["x"] = -1
                mouse_pos["y"] = -1
    return True
```

In [45]:
# Codi de l'exemple
import utils; utils.show_code('./exemple006.py', function_name='app_run')

```python
def app_run():
    global mouse_pos, eye_left, eye_right

    eye_left["x"] = mouse_pos["x"] - 10
    if (eye_left["x"] < 250):
        eye_left["x"] = 250
    elif (eye_left["x"] > 280):
        eye_left["x"] = 280

    eye_left["y"] = mouse_pos["y"] - 10
    if (eye_left["y"] < 150):
        eye_left["y"] = 150
    elif (eye_left["y"] > 205):
        eye_left["y"] = 205

    eye_right["x"] = mouse_pos["x"] - 10
    eye_right["y"] = mouse_pos["y"] - 10
```

In [ ]:
# Codi de l'exemple
import utils; utils.show_code('./exemple006.py', function_name='app_draw')

In [46]:
# Codi de l'exemple
import utils; utils.run_code('./exemple006.py')

### Exercici 004

Al codi **"exercici004.py"**, arregla l'exemple anterior perquè l'ull dret estigui dins dels seus limits

In [47]:
# Fer anar el codi de l'exercici
import utils; utils.run_code('./exercici004.py')

## Botons del ratolí apretats o alliberats

Al igual que amb el teclat, es produeix un event quan s'apreta o allibera un botó del ratolí.

Amb aquest event, podem saber les coordenades on s'ha fet 'click'.

In [49]:
# Codi de l'exemple
import utils; utils.show_code('./exemple007.py', function_name='app_events')

```python
def app_events():
    global mouse_pos, mouse_down
    mouse_inside = pygame.mouse.get_focused() # El ratolí està dins de la finestra?

    for event in pygame.event.get():
        if event.type == pygame.QUIT: # Botó tancar finestra
            return False
        elif event.type == pygame.MOUSEMOTION:
            if mouse_inside:
                mouse_pos["x"] = event.pos[0]
                mouse_pos["y"] = event.pos[1]
            else:
                mouse_pos["x"] = -1
                mouse_pos["y"] = -1
        elif event.type == pygame.MOUSEBUTTONDOWN:
            mouse_down = True
        elif event.type == pygame.MOUSEBUTTONUP:
            mouse_down = False
    return True
```

In [50]:
# Codi de l'exemple
import utils; utils.show_code('./exemple007.py', function_name='app_run')

```python
def app_run():
    global mouse_pos, mouse_down, square_clicked, circle_clicked

    if mouse_down:
        if (mouse_pos["x"] > 100 and 
            mouse_pos["y"] > 150 and 
            mouse_pos["x"] < (100 + 200) and
            mouse_pos["y"] < (150 + 50)):
                square_clicked = True

        center = { "x": 400, "y": 175 }
        if is_point_in_circle(mouse_pos, center, 50):
                circle_clicked = True

    else:
        square_clicked = False
        circle_clicked = False
```

In [51]:
# Codi de l'exemple
import utils; utils.show_code('./exemple007.py', function_name='app_draw')

```python
def app_draw():
    global square_clicked, circle_clicked

    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella
    utils.draw_grid(pygame, screen, 50)

    # Dibuixar text d'ajuda
    font = pygame.font.SysFont("Arial", 24)
    text = font.render('Fes click als objectes', True, BLACK)
    screen.blit(text, (50, 50))

    # Quadre i cercle
    color = BLACK
    if square_clicked:
        color = BLUE
    pygame.draw.rect(screen, color, pygame.Rect(100, 150, 200, 50))

    color = BLACK
    if circle_clicked:
        color = GREEN
    pygame.draw.circle(screen, color, (400, 175), 50)

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

In [52]:
# Codi de l'exemple
import utils; utils.run_code('./exemple007.py')

## Arrossegar objectes

Per arrossegar s'han de seguir diversos passos:

- Capturar els events de botons de mouse
- Comprovar si es fa click a sobre d'un objecte o a fora
- Apuntar la diferència entre la posició on es fa click i l'origen de dibuix de l'objecte
- Moure l'objecte amb el ratolí (segons la diferència anterior)
- Alliberar l'arrosegament quan s'allibera el botó del ratolí

**Nota:** Per guardar la diferència entre la posició on s'ha fet click i l'origen del dibuix, tenim una variable **"drag_offset"**



In [54]:
# Codi de l'exemple
import utils; utils.show_code('./exemple008.py', function_name='app_events')

```python
def app_events():
    global mouse_pos, mouse_down
    mouse_inside = pygame.mouse.get_focused()  # El ratolí està dins de la finestra?

    for event in pygame.event.get():
        if event.type == pygame.QUIT:  # Botó tancar finestra
            return False
        elif event.type == pygame.MOUSEMOTION:
            if mouse_inside:
                mouse_pos["x"], mouse_pos["y"] = event.pos
        elif event.type == pygame.MOUSEBUTTONDOWN:
            mouse_down = True
        elif event.type == pygame.MOUSEBUTTONUP:
            mouse_down = False
    return True
```

In [55]:
# Codi de l'exemple
import utils; utils.show_code('./exemple008.py', function_name='app_run')

```python
def app_run():
    global rectangle_pos, circle_center, square_dragging, circle_dragging, drag_offset

    # Inici de l'arrossegament en fer clic dins del rectangle o cercle
    if mouse_down:
        if not square_dragging and not circle_dragging:  # Només detecta al començar l'arrossegament
            if rectangle_pos.collidepoint(mouse_pos["x"], mouse_pos["y"]):
                square_dragging = True
                drag_offset["x"] = mouse_pos["x"] - rectangle_pos.x
                drag_offset["y"] = mouse_pos["y"] - rectangle_pos.y
            elif is_point_in_circle(mouse_pos, circle_center, circle_radius):
                circle_dragging = True
                drag_offset["x"] = mouse_pos["x"] - circle_center["x"]
                drag_offset["y"] = mouse_pos["y"] - circle_center["y"]
    else:
        # Alliberar l'arrossegament quan s'aixeca el ratolí
        square_dragging = False
        circle_dragging = False

    # Actualitza la posició del rectangle si es fa "drag"
    if square_dragging:
        rectangle_pos.x = mouse_pos["x"] - drag_offset["x"]
        rectangle_pos.y = mouse_pos["y"] - drag_offset["y"]

    # Actualitza la posició del cercle si es fa "drag"
    if circle_dragging:
        circle_center["x"] = mouse_pos["x"] - drag_offset["x"]
        circle_center["y"] = mouse_pos["y"] - drag_offset["y"]
```

In [56]:
# Codi de l'exemple
import utils; utils.show_code('./exemple008.py', function_name='app_draw')

```python
def app_draw():
    # Pintar el fons de blanc
    screen.fill(WHITE)

    # Dibuixar la graella
    utils.draw_grid(pygame, screen, 50)

    # Dibuixar el rectangle
    rectangle_color = BLUE if square_dragging else BLACK
    pygame.draw.rect(screen, rectangle_color, rectangle_pos)

    # Dibuixar el cercle
    circle_color = GREEN if circle_dragging else BLACK
    pygame.draw.circle(screen, circle_color, (circle_center["x"], circle_center["y"]), circle_radius)

    # Actualitzar el dibuix a la finestra
    pygame.display.update()
```

In [53]:
# Codi de l'exemple
import utils; utils.run_code('./exemple008.py')

### Exercici 005

Al codi **"exercici005.py"**, modifica l'exemple anterior per tal que quan els dos poligons estàn sobreposats i a més s'estàn arrossegant, es dibuixin tots dos de color taronja.

In [129]:
# Fer anar el codi de l'exercici
import utils; utils.run_code('./exercici005.py')

### Exercici 006

Al codi **"exercici006.py"**, modifica el codi per tal fer fer el següent programa:

- El programa té una matriu "mat" de 10x10 on cada casella té un número del 0 al 10 escollits a l'atzar.

- La matriu anterior s'ha de dibuixar com un tauler de 10x10 on cada cel·la és de mida 50x50, i a dins té el número escollit.

- Quan el mouse fa click a dins d'una cel·la del tauler, la cel·la s'emplena de color verd clar com a fons, a sobre continua mostrant el número de la matiu que li correspon.

- Si la cel·la on s'ha fet click té el mateix número en una cel·la contígua en forma de creu, els números iguals s'esborren.